core/iwasm/aot/aot_runtime.c

```diff
@@ -1444,8 +1465,26 @@ aot_call_function(WASMExecEnv *exec_env, AOTFunctionInstance *function,
         }
 #endif
 
-        ret = invoke_native_internal(exec_env, function->u.func.func_ptr,
-                                     func_type, NULL, NULL, argv, argc, argv);
+#ifdef OS_ENABLE_BLOCK_INSN_INTERRUPT
+        wasm_exec_env_push_jmpbuf(exec_env, &jmpbuf_node);
+        wasm_runtime_set_exec_env_tls(exec_env);
+        if (os_setjmp(jmpbuf_node.jmpbuf) == 0) {
+#endif
+            ret =
+                invoke_native_internal(exec_env, function->u.func.func_ptr,
+                                       func_type, NULL, NULL, argv, argc, argv);
+#ifdef OS_ENABLE_BLOCK_INSN_INTERRUPT
+        }
+        else {
+            ret = false;
```

Comment:
who frees eg. "argv1" in wasm_runtime_invoke_native on a termination?

while OS_ENABLE_HW_BOUND_CHECK seems to have the same problem,
it seems much more complex to fix for OS_ENABLE_BLOCK_INSN_INTERRUPT because it's an async event.

Reply:
Yes, this is a potential issue, maybe there are two ways to resolve it: (1) avoid allocating memory in wasm_runtime_invoke_native, just define a large enough buffer for argv, (2) add a list in exec_env, add the allocating the memory to the list, and free it later.

Reply:
(2) needs to block the signal for OS_ENABLE_BLOCK_INSN_INTERRUPT.
maybe we should invent a convenient api for common cases like memory allocation.

Reply:
Are you referring to the call at L1396? Isn't it freed at L1458 as before my changes?

Reply:
I made a try with valgrind using interpreter and AOT, with or without HW bound check and no memory leakage is detected. Unfortunately, in case of AOT I see many warnings about invalid memory accesses that may explain the weird behavior that I reported in another comment.

Reply:
> Are you referring to the call at L1396? Isn't it freed at L1458 as before my changes?

i'm not sure which file and version your line numbers are for.

i meant eg.
https://github.com/bytecodealliance/wasm-micro-runtime/blob/2c8375ee67964a4ef99dde2faa1c5225051adca0/core/iwasm/common/wasm_runtime_common.c#L3506
https://github.com/bytecodealliance/wasm-micro-runtime/blob/2c8375ee67964a4ef99dde2faa1c5225051adca0/core/iwasm/common/wasm_runtime_common.c#L3790

when it escapes with a long jump, it might not be executed.

Reply:
Got it, for this one https://github.com/bytecodealliance/wasm-micro-runtime/blob/2c8375ee67964a4ef99dde2faa1c5225051adca0/core/iwasm/common/wasm_runtime_common.c#L4130 in my tests the `if` is never entered and so there's no memory allocation, that's why I was not observing any memory leakage in my configuration.

But, right, there's a potential memory leakage.

Reply:
(3) what if we allocate it in the parent function?

Reply:
@yamt How about limiting the max argument count to 32 and reporting error if argument count is larger than that? So there is no memory allocation in wasm_runtime_invoke_native.
I think in normal usage, there isn't a function taking so many arguments, 32 is big enough: `uint64 argv_buf[32] = { 0 }`

Reply:
Have a test on the spec cases, need more size, considering 36 or 48

Reply:
well, `argv1` is just an example. it isn't the only problem or the biggest problem.
the native functions themselves likely have more important things to cleanup.
eg. locks used in wasi implementation need to be released when it's terminated with the signal.

the only simple solution i can think of right now involves block/unblock of the signal.
like:
```
block signal
get a resource (allocate memory / acquire lock / etc)
register cleanup handler for the resource
unblock signal
```
how do you think?

Reply:
The issue is only introduced by this PR which may raise the signal to kill other threads, right?

Reply:
right.
well, there might be similar issues in the hw bound check code. i dunno. but at least this PR makes far much code affected because the signal in question is sent by other threads asynchronously.

Reply:
It is really complex, we just don't know the possible behaviors of user implemented native API, there are some questions:
- How many resources do we want to protect? The memory allocated from wasm_runtime_malloc, the initialized lock/cond? And if developer allocates memory from non wasm_runtime_malloc?
- What the blocking situations include? `while (1)` in native, sleep, timedwait, and lock? Is it reasonable to terminate them?
- How to protect the resources, the process may be like what you proposed, but need to break down them. Do we want to implement them in this PR, or in another PR?

Reply:
If we decide to implement it in another PR we can probably disable the feature by default.

Reply:
Yes, I think we had better create another PR for the further enhancement. How about we creating another branch `dev/interrup_block_insn` from main, submitting this PR to that branch, and then we can submit following PRs to it. And after the feature is tested more and is relatively stable, we can merge it to main.
@yamt What's you opinion?

Reply:
i'm afraid that it needs to be more like code audit than testing.
i suppose disabling it by default and enabling it on per-function basis can help.

i feel having another branch is too much as far as it's disabled by default. but i don't have strong opinions either way.

Reply:
Should we focus on avoiding memory allocations in the function called after `os_setjmp`? E.g. 
https://github.com/eloparco/wasm-micro-runtime/blob/9d6dc29a2d4dc85cf759dc3066741b2a832423f3/core/iwasm/interpreter/wasm_runtime.c#L2142
and
https://github.com/eloparco/wasm-micro-runtime/blob/9d6dc29a2d4dc85cf759dc3066741b2a832423f3/core/iwasm/interpreter/wasm_runtime.c#L2144

Because if a signal is received before `os_setjmp`, it will be ignored in the signal handler. If it's received after, [this](https://github.com/eloparco/wasm-micro-runtime/blob/9d6dc29a2d4dc85cf759dc3066741b2a832423f3/core/iwasm/interpreter/wasm_runtime.c#L2156) is the function that could leak memory. 

After https://github.com/eloparco/wasm-micro-runtime/blob/9d6dc29a2d4dc85cf759dc3066741b2a832423f3/core/iwasm/interpreter/wasm_runtime.c#L2160, there is no `jmpbuf`, so the signal handler would [return](https://github.com/eloparco/wasm-micro-runtime/blob/9d6dc29a2d4dc85cf759dc3066741b2a832423f3/core/iwasm/common/wasm_runtime_common.c#L330) without jumping.

Reply:
> i'm afraid that it needs to be more like code audit than testing. i suppose disabling it by default and enabling it on per-function basis can help.
> 
> i feel having another branch is too much as far as it's disabled by default. but i don't have strong opinions either way.

Two reasons to merge it into a development branch:
(1) avoid merging a not so mature PR to the main branch quickly, the PR changes the vmcore part, I think we had better be cautious and do more test on it. And actually the feature is enabled by default for posix platforms, see https://github.com/bytecodealliance/wasm-micro-runtime/pull/1930/files#diff-77a81caccc9f723fd431bfc253006da5129f55a3c93b6275b01ec1f642f9b6ee
```C
#if WASM_DISABLE_BLOCK_INSN_INTERRUPT == 0 && WASM_ENABLE_THREAD_MGR != 0
#define OS_ENABLE_BLOCK_INSN_INTERRUPT
```
(2) we can then try to resolve the memory allocation issue: protect the allocated memory after os_setjmp, and disable the signal before/after calling wasm_runtime_malloc/free.

I will merge it to main branch as soon as possible when (2) is done and well tested.

Reply:
> Should we focus on avoiding memory allocations in the function called after `os_setjmp`?

Agree, let's resolve the memory allocation issue firstly.

Reply:
> I will merge it to main branch as soon as possible when (2) is done and well tested.

At this point, we can probably just have the feature disabled by default until it's well tested.

> the only simple solution i can think of right now involves block/unblock of the signal.

The tricky part is that it's not always possible to just block signals before the malloc and unblock after the free, because the blocking instruction can be in the middle. Apart from `argv1 ` that we could avoid allocating dynamically, the problem is in the native function calls.

Not sure what's the best solution there, maybe the approach of keeping a list in `exec_env` is not that bad, but only for native calls otherwise we only release the entire allocated memory at the end of the execution. But then how do we decide when to use the exec_env list and when to free normally? It seems quite error prone.

Reply:
Yes, suggest to add a list of allocated memories after setjmp in the exec_env.
@eloparco I have created branch `dev/interrupt_block_insn`, could you upload another patch to it? Or do you mind I refine some code and upload the PR by myself?

Reply:
Thanks @wenyongh, I pushed the changes I made so far in #1946.

Reply:
@eloparco I submitted PR #1947 to refactor the feature, please help review, thanks.

Reply:
Sorry, the source branch and dest branch are invalid, closed it and submitted #1948 instead